### PR TITLE
intern strings extracted from small dictionaries

### PR DESCRIPTION
**Pull request by richardstartin**
`StringDictionary.getStringValue` shows up in most profiles I have seen of Pinot. It used to be bottlenecked on finding the null terminator in `FixedByteValueReaderWriter.getUnpaddedString` but this was sped up in #7708, now the number of allocations is the bottleneck.

This PR adds an interning table in `StringDictionary` which is only used when the size of the strings (excluding object headers and shallow layout) to be interned is guaranteed to be less than 10MB, which is roughly equivalent to the block sizes used by `TransformFunction`s. When the dictionary fits into the intern table, this improves throughput (over 10x in some cases) eliminates allocation and GC time. When the dictionary is too large to be interned within 10MB, throughput and allocation rate do not regress.

Without interning table
```
Benchmark                                                                                   (_length)  (_nativeOrder)  (_paddingByte)  (_values)  Mode  Cnt           Score           Error   Units
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8            true              42        512  avgt    5          14.608 ±         0.112   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8            true              42        512  avgt    5       36304.396 ±         2.842    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                              8            true              42        512  avgt    5          28.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8            true              42       4096  avgt    5         139.851 ±         0.646   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8            true              42       4096  avgt    5      290683.783 ±        27.218    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                              8            true              42       4096  avgt    5          25.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8            true              42       8192  avgt    5         287.664 ±         9.572   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8            true              42       8192  avgt    5      581535.774 ±        55.799    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                              8            true              42       8192  avgt    5          24.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8            true              42      16384  avgt    5         566.878 ±        14.509   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8            true              42      16384  avgt    5     1162791.378 ±       110.594    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                              8            true              42      16384  avgt    5          24.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8           false              42        512  avgt    5          13.909 ±         0.202   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8           false              42        512  avgt    5       36304.378 ±         2.709    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                              8           false              42        512  avgt    5          30.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8           false              42       4096  avgt    5         141.954 ±         4.906   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8           false              42       4096  avgt    5      290683.864 ±        27.699    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                              8           false              42       4096  avgt    5          23.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8           false              42       8192  avgt    5         282.691 ±        12.511   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8           false              42       8192  avgt    5      581535.614 ±        54.560    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                              8           false              42       8192  avgt    5          23.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8           false              42      16384  avgt    5         565.192 ±        20.916   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8           false              42      16384  avgt    5     1162791.245 ±       109.161    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                              8           false              42      16384  avgt    5          25.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32            true              42        512  avgt    5          21.390 ±         0.539   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32            true              42        512  avgt    5       54672.574 ±         4.117    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                             32            true              42        512  avgt    5          28.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32            true              42       4096  avgt    5         209.035 ±        13.126   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32            true              42       4096  avgt    5      438077.624 ±        40.415    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                             32            true              42       4096  avgt    5          24.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32            true              42       8192  avgt    5         400.035 ±        12.125   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32            true              42       8192  avgt    5      875754.949 ±        78.816    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                             32            true              42       8192  avgt    5          25.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32            true              42      16384  avgt    5         818.458 ±        12.627   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32            true              42      16384  avgt    5     1754046.124 ±       158.981    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                             32            true              42      16384  avgt    5          24.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32           false              42        512  avgt    5          22.899 ±         0.597   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32           false              42        512  avgt    5       54672.625 ±         4.468    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                             32           false              42        512  avgt    5          26.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32           false              42       4096  avgt    5         204.732 ±         8.169   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32           false              42       4096  avgt    5      438077.533 ±        39.776    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                             32           false              42       4096  avgt    5          25.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32           false              42       8192  avgt    5         402.713 ±        10.173   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32           false              42       8192  avgt    5      875754.719 ±        78.094    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                             32           false              42       8192  avgt    5          26.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32           false              42      16384  avgt    5         897.035 ±         9.160   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32           false              42      16384  avgt    5     1754047.764 ±       174.105    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                             32           false              42      16384  avgt    5          24.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024            true              42        512  avgt    5         142.570 ±        10.162   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024            true              42        512  avgt    5      808595.898 ±        28.537    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                           1024            true              42        512  avgt    5          36.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024            true              42       4096  avgt    5        1184.816 ±        29.738   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024            true              42       4096  avgt    5     6511319.379 ±       229.643    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                           1024            true              42       4096  avgt    5          54.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024            true              42       8192  avgt    5        2500.836 ±        95.902   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024            true              42       8192  avgt    5    13036130.134 ±       482.732    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                           1024            true              42       8192  avgt    5          52.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024            true              42      16384  avgt    5        5789.566 ±       450.470   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024            true              42      16384  avgt    5    26110782.836 ±      1100.930    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                           1024            true              42      16384  avgt    5          38.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024           false              42        512  avgt    5         149.560 ±         2.794   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024           false              42        512  avgt    5      808595.962 ±        28.838    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                           1024           false              42        512  avgt    5          54.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024           false              42       4096  avgt    5        1248.535 ±        38.234   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024           false              42       4096  avgt    5     6511321.162 ±       242.151    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                           1024           false              42       4096  avgt    5          52.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024           false              42       8192  avgt    5        2657.626 ±       221.590   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024           false              42       8192  avgt    5    13036133.971 ±       511.855    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                           1024           false              42       8192  avgt    5          48.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024           false              42      16384  avgt    5        5959.743 ±       544.894   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024           false              42      16384  avgt    5    26110794.161 ±      1192.076    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                           1024           false              42      16384  avgt    5          38.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536            true              42        512  avgt    5       10276.808 ±       116.498   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536            true              42        512  avgt    5    50814880.376 ±      1994.851    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536            true              42        512  avgt    5          51.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536            true              42       4096  avgt    5       81733.900 ±       199.826   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536            true              42       4096  avgt    5   405690051.938 ±     15044.988    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536            true              42       4096  avgt    5          53.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536            true              42       8192  avgt    5      163259.840 ±      9478.884   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536            true              42       8192  avgt    5   809156906.286 ±     27907.233    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536            true              42       8192  avgt    5          55.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536            true              42      16384  avgt    5      321912.650 ±      4214.615   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536            true              42      16384  avgt    5  1615953222.800 ±     48827.319    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536            true              42      16384  avgt    5          65.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536           false              42        512  avgt    5       10318.616 ±       163.180   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536           false              42        512  avgt    5    50790311.946 ±      2014.396    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536           false              42        512  avgt    5          50.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536           false              42       4096  avgt    5       81945.555 ±      1225.584   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536           false              42       4096  avgt    5   405493485.785 ±     15044.999    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536           false              42       4096  avgt    5          54.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536           false              42       8192  avgt    5      163695.100 ±     10944.886   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536           false              42       8192  avgt    5   808764318.743 ±     32678.816    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536           false              42       8192  avgt    5          56.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536           false              42      16384  avgt    5      322787.764 ±      4244.671   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536           false              42      16384  avgt    5  1615166949.200 ±     48882.435    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536           false              42      16384  avgt    5          65.000                      ms
```

With interning table
```
Benchmark                                                                                   (_length)  (_nativeOrder)  (_paddingByte)  (_values)  Mode  Cnt           Score           Error   Units
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8            true              42        512  avgt    5           1.297 ±         0.024   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8            true              42        512  avgt    5           0.035 ±         0.251    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8            true              42       4096  avgt    5          15.779 ±         0.477   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8            true              42       4096  avgt    5           0.427 ±         3.063    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8            true              42       8192  avgt    5          31.753 ±         0.791   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8            true              42       8192  avgt    5           0.858 ±         6.148    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8            true              42      16384  avgt    5          63.508 ±         5.653   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8            true              42      16384  avgt    5           1.747 ±        12.555    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8           false              42        512  avgt    5           1.320 ±         0.066   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8           false              42        512  avgt    5           0.036 ±         0.256    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8           false              42       4096  avgt    5          15.698 ±         0.293   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8           false              42       4096  avgt    5           0.428 ±         3.077    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8           false              42       8192  avgt    5          25.852 ±         0.249   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8           false              42       8192  avgt    5           0.701 ±         5.031    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                       8           false              42      16384  avgt    5          68.022 ±         4.073   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                   8           false              42      16384  avgt    5           1.836 ±        13.137    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32            true              42        512  avgt    5           1.289 ±         0.045   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32            true              42        512  avgt    5           0.035 ±         0.249    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32            true              42       4096  avgt    5          15.847 ±         0.650   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32            true              42       4096  avgt    5           0.425 ±         3.039    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32            true              42       8192  avgt    5          33.847 ±         2.629   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32            true              42       8192  avgt    5           0.942 ±         6.795    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32            true              42      16384  avgt    5          62.715 ±         5.622   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32            true              42      16384  avgt    5           1.718 ±        12.295    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32           false              42        512  avgt    5           1.322 ±         0.112   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32           false              42        512  avgt    5           0.035 ±         0.250    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32           false              42       4096  avgt    5          16.403 ±         1.556   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32           false              42       4096  avgt    5           0.445 ±         3.185    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32           false              42       8192  avgt    5          32.329 ±         3.328   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32           false              42       8192  avgt    5           0.889 ±         6.488    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                      32           false              42      16384  avgt    5          61.608 ±         6.302   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                  32           false              42      16384  avgt    5           1.603 ±        11.540    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024            true              42        512  avgt    5           1.315 ±         0.057   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024            true              42        512  avgt    5           0.035 ±         0.253    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024            true              42       4096  avgt    5          14.853 ±         0.417   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024            true              42       4096  avgt    5           0.393 ±         2.850    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024            true              42       8192  avgt    5          31.525 ±         0.863   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024            true              42       8192  avgt    5           0.841 ±         6.090    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024            true              42      16384  avgt    5        5768.363 ±       469.567   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024            true              42      16384  avgt    5    26110781.716 ±      1090.873    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                           1024            true              42      16384  avgt    5          38.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024           false              42        512  avgt    5           1.306 ±         0.069   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024           false              42        512  avgt    5           0.035 ±         0.250    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024           false              42       4096  avgt    5          14.877 ±         0.339   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024           false              42       4096  avgt    5           0.395 ±         2.868    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024           false              42       8192  avgt    5          29.733 ±         1.415   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024           false              42       8192  avgt    5           0.791 ±         5.741    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                    1024           false              42      16384  avgt    5        5758.051 ±       607.162   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm                1024           false              42      16384  avgt    5    26110780.623 ±      1077.232    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                           1024           false              42      16384  avgt    5          40.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536            true              42        512  avgt    5       13296.618 ±      3087.935   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536            true              42        512  avgt    5    50814968.906 ±      2568.730    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536            true              42        512  avgt    5          33.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536            true              42       4096  avgt    5       90701.280 ±     50476.410   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536            true              42       4096  avgt    5   405690141.832 ±     14932.905    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536            true              42       4096  avgt    5          49.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536            true              42       8192  avgt    5      164604.913 ±      9421.941   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536            true              42       8192  avgt    5   809157004.838 ±     27892.128    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536            true              42       8192  avgt    5          56.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536            true              42      16384  avgt    5      324320.649 ±     25571.625   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536            true              42      16384  avgt    5  1615955323.733 ±     65470.493    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536            true              42      16384  avgt    5          61.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536           false              42        512  avgt    5       10276.230 ±       258.071   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536           false              42        512  avgt    5    50790310.717 ±      2016.330    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536           false              42        512  avgt    5          51.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536           false              42       4096  avgt    5       81751.354 ±      1415.242   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536           false              42       4096  avgt    5   405493480.369 ±     15024.848    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536           false              42       4096  avgt    5          53.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536           false              42       8192  avgt    5      162787.523 ±      5507.412   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536           false              42       8192  avgt    5   808763775.314 ±     27871.817    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536           false              42       8192  avgt    5          57.000                      ms
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary                                   65536           false              42      16384  avgt    5      322235.766 ±      7143.305   us/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.alloc.rate.norm               65536           false              42      16384  avgt    5  1615166960.400 ±     48823.866    B/op
BenchmarkFixedByteValueReaderWriter.readStringsFromDictionary:·gc.time                          65536           false              42      16384  avgt    5          65.000                      ms
```

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/StringDictionary.java [hunk 1]
```diff
@@ -41,84 +47,117 @@ public DataType getValueType() {
...
+      _internTable[dictId] = interned;
+    }
+    return interned;
+  }
+
+  @Override
+  public void close()
+      throws IOException {
+    if (_internTable != null) {
+      Arrays.fill(_internTable, null);
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/StringDictionary.java:159]: Is this required? If so, does setting `_internTable = null` have better performance?

Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/StringDictionary.java:159]: I don’t think we need this, but note that the intern table is final. I’ll just remove it.

**Comment by richardstartin** [comment]
> High level question: the improvement in this PR is very similar to the `OnHeapStringDictionary`. I am thinking maybe we should just use `OnHeapStringDictionary` if the dictionary size is small

I think it’s worth comparison. This has a couple of advantages:
- fewer types, making it more likely the calls to getStringValue inline into hot loops, if they don’t inline it affects what can be done with that loop.
- Only store the values which actually get unpacked

**Comment by richardstartin** [comment]
Looking at `OnHeapStringDictionary`, it's much heavier than an intern table:

```java
  private final String[] _unpaddedStrings;
  private final Object2IntOpenHashMap<String> _unPaddedStringToIdMap;
  private final String[] _paddedStrings;
```

The padded strings and string to id map don't serve much purpose on the hot path.

**Comment by Jackie-Jiang** [comment]
The `_paddedStrings` should not be setup in normal case (it's for backward-compatibility for some very old segment format). The map is used for looking up dict id, which is used for filtering phase (less frequently called comparing to read value from dict id).
After a second thought, I feel adding this on-heap array to the default dictionary can potentially cause problems. Say a server has 10K segments loaded, each segment has one 10M string dictionary, it could consume 100G heap memory in the worst case

**Comment by richardstartin** [comment]
Yes, and many of the strings will be duplicated across many of those 10K segments too. You illustrate well how much memory bandwidth the query layer requires, because each transform function in each query for each segment will construct a similarly sized array. I think it would be better to introduce a `StringView` type which refers to the bytes without needing to copy them.

**Pull request closed without merging**